⚠ Remove deprecated Defaulter and Validator interfaces

troy0820 opened the pull request:
We deprecated the `admission.Validator` and the `admission.Defaulter` interfaces and these are said to be removed in ~~release-0.19.~~ release-0.20  

 Closes #2641

- [x] Remove deprecated interfaces
- [x] ~~Remove tests for those types~~ Change tests for the deprecated types to the CustomDefaulter/CustomValidator
- [x] ~~Change~~ Remove log message that point to the now removed interfaces for the builder.
- [x] Moved type `Warnings` to `validator_custom.go`

sbueringer commented:
@troy0820 can you remind me where we said we're going to remove them with 0.19?

IIRC it was on some issue, can you please link it in the PR description?

That being said, not sure when we should do it. I think this will require a lot of folks to make changes, maybe we should give them more time. I don't think we have anything pressing that requires us to drop them very soon.

@vincepri @alvaroaleman wdyt?

troy0820 commented:
> can you remind me where we said we're going to remove them with 0.19?
IIRC it was on some issue, can you please link it in the PR description?

Updated the PR description with the issue. 

> That being said, not sure when we should do it. I think this will require a lot of folks to make changes, maybe we should give them more time. I don't think we have anything pressing that requires us to drop them very soon.

I’m not in any rush to remove them but it was put against the v0.19 milestone. We can wait if the changes need to be made other places as well as give people more time.

sbueringer commented:
@troy0820 Thx for working on this! I talked to @alvaroaleman @vincepri and we would merge it right after the 0.19.0 release so it would be part of 0.20.0. I'll review the PR soon

sbueringer commented on pkg/builder/webhook_test.go:
```diff
@@ -77,138 +77,6 @@ func runTests(admissionReviewVersion string) {
-		ExpectWithOffset(1, w.Code).To(Equal(http.StatusNotFound))
-	})
-
-	It("should scaffold a defaulting webhook which recovers from panics", func() {
```
Let's please keep test coverage for recover panics (and change this test to use customdefaulter)

sbueringer reviewed (commented):
Looks good overall. Mostly just wanting to make sure that for whatever coverage we drop we'll have the corresponding coverage with the new types

camilamacedo86 commented:
Could we not merge this one for the next release?
We need ensure that those changes will be applied in the Kubebuilder scaffolds beforehand.

sbueringer commented:
> Hi folks,
> 
> Could we please not merge this one for the next release? We need ensure that those changes will be applied in the Kubebuilder scaffolds beforehand.
> 
> @sbueringer @grzesuav
> 
> /hold

We are not going to merge it for 0.19 (August), but we are going to merge it for 0.20 (December/January).

We had this deprecated since 0.17, I don't think we should delay it further than December

sbueringer commented:
/hold cancel

sbueringer commented:
@troy0820 Do you have time to work on the open findings?

troy0820 commented:
Yes, I should have them finish this week.  @sbueringer

troy0820 commented on pkg/webhook/admission/defaulter_test.go:
```diff
@@ -1,68 +0,0 @@
-
-var _ = Describe("Defaulter Handler", func() {
-
-	It("should return ok if received delete verb in defaulter handler", func() {
```
Do you want this to run through the `webhook_test.go` in the admissions package or create this in the `pkg/builder/webhook_test.go` which can test the whole flow? @sbueringer 

We currently do have any coverage for custom validator/defaulter in this package.

sbueringer replied:
> We currently do have any coverage for custom validator/defaulter in this package.

> We currently do **not** have any coverage for custom validator/defaulter in this package.

Right?

The more I'm thinking about it the more I think we should have these tests in defaulter_custom_test.go to actually *unit* test this code (same for validator).

(to be clear I would just move the existing tests in defaulter_test.go / validator_test.go over to the corresponing defaulter_custom_test.go / .. files. Let's keep webhook_test.go tests where they are)

sbueringer replied:
@troy0820 same for the tests from validator_test.go (sorry not sure if that's WIP right now :))

troy0820 replied:
Indeed, a little more work for the validator_test.go one.  Should be coming soon.

troy0820 commented:
@sbueringer addressed the comments with the thumbs up but still have a question around this from above:
> Do you want this to run through the `webhook_test.go` in the admissions package or create this in the `pkg/builder/webhook_test.go` which can test the whole flow? @sbueringer
> 
> We currently do have any coverage for custom validator/defaulter in this package.

Also not squashed so the diff can be visualized within the context of the comments.

troy0820 commented on pkg/builder/webhook_test.go:
```diff
@@ -660,24 +474,6 @@ func runTests(admissionReviewVersion string) {
 		ExpectWithOffset(1, w.Body).To(ContainSubstring(`"code":200`))
 	})
-
-	It("should send an error when trying to register a webhook with more than one For", func() {
-		By("creating a controller manager")
-		m, err := manager.New(cfg, manager.Options{})
-		ExpectWithOffset(1, err).NotTo(HaveOccurred())
-
-		By("registering the type in the Scheme")
-		builder := scheme.Builder{GroupVersion: testDefaulterGVK.GroupVersion()}
-		builder.Register(&TestDefaulter{}, &TestDefaulterList{})
-		err = builder.AddToScheme(m.GetScheme())
-		ExpectWithOffset(1, err).NotTo(HaveOccurred())
-
-		err = WebhookManagedBy(m).
-			For(&TestDefaulter{}).
-			For(&TestDefaulter{}).
-			Complete()
-		Expect(err).To(HaveOccurred())
-	})
```
This test is lost but I can add it back as the For, For should fail for CustomDefaulter/CustomValidator as well.

troy0820 replied:
Nevermind, this got moved up to line 364

troy0820 commented on pkg/webhook/admission/validator_custom_test.go:
```diff
@@ -448,11 +459,11 @@ var _ = Describe("validatingHandler", func() {
 })
 
 // fakeValidator provides fake validating webhook functionality for testing
```
```suggestion
// fakeCustomValidator provides fake validating webhook functionality for testing
```

sbueringer commented on pkg/builder/webhook_test.go:
```diff
@@ -419,157 +361,25 @@ func runTests(admissionReviewVersion string) {
 		ExpectWithOffset(1, w.Body).To(ContainSubstring(`"message":"panic: fake panic test [recovered]`))
 	})
 
-	It("should scaffold a validating webhook with a custom validator", func() {
```
Is this test case lost?

Sorry it's really had to diff. Can you please check the diff and change the ordering in your new version a bit to get a minimal diff? (it's really had because it looks like some test cases were moved around)

troy0820 replied:
Yeah this moved to line 219.  I can try to minimize the diff but deleting tests are going to blob where it's red.  I'll see what I can do.

sbueringer replied:
Yup that's fine. I'm just currently shuffling tests around to see what the actual diff is :D

sbueringer replied:
Sounds good, no worries. Thx!

sbueringer replied:
Reviewed, thank you very much! A lot easier to tell now that everything is fine :)

sbueringer reviewed (commented):
Thx, last one from my side

sbueringer commented:
/lgtm

/assign @alvaroaleman @vincepri

sbueringer commented:
/approve

alculquicondor commented on pkg/builder/webhook_test.go:
```diff
@@ -77,7 +77,7 @@ func runTests(admissionReviewVersion string) {
 	})
 
-	It("should scaffold a defaulting webhook if the type implements the Defaulter interface", func() {
+	It("should scaffold a custom defaulting webhook if the type implements the CustomDefaulter interface", func() {
```
why didn't you remove this test?

There is already `should scaffold a defaulting webhook with a custom defaulter` below.

sbueringer replied:
I think this was an oversight, we can delete this one

sbueringer replied:
I think same for `should scaffold a custom validating webhook if the type implements the CustomValidator interface`
